[Refactor] Split up inheritance structure of TransformerEncoder and BERTEncoder

leezu opened the pull request:
## Description ##
Split up inheritance structure of TransformerEncoder and BERTEncoder

## Checklist ##
### Essentials ###
- [X] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Split up inheritance structure of TransformerEncoder and BERTEncoder

## Comments ##
- Removes `BERTLayerNorm` and `BERTPositionwiseFFN`
- Changes the `layer_norm_eps=None` default to `layer_norm_eps=1e-5` for Transformer models and `layer_norm_eps=1e-12` for BERT models. The distinction was previously done by using `BERTLayerNorm` vs `nn.LayerNorm`.

cc @dmlc/gluon-nlp-team @fierceX

eric-haibin-lin commented on src/gluonnlp/model/bert.py:
```diff
@@ -167,40 +96,158 @@ class BERTEncoder(BaseTransformerEncoder):
+                    activation=activation, layer_norm_eps=layer_norm_eps)
+                self.transformer_cells.add(cell)
+
+    def __call__(self, inputs, states=None, valid_length=None): #pylint: disable=arguments-differ
```
My comment is not related to refactoring: is there any issue using the recently added None support in gluon, and removing `__call__`? Just curious

leezu replied:
Gluon doesn't support keyword arguments. So removing call will break API. We should add keyword argument support to Gluon..

eric-haibin-lin commented on src/gluonnlp/model/bert.py:
```diff
@@ -342,12 +417,11 @@ def __init__(self, encoder, vocab_size=None, token_type_vocab_size=None, units=N
                  word_embed=None, token_type_embed=None, use_pooler=True, use_decoder=True,
                  use_classifier=True, use_token_type_embed=True, prefix=None, params=None):
-        super(BERTModel, self).__init__(prefix=prefix, params=params)
+        super().__init__(prefix=prefix, params=params)
```
Stupid question: is not specifying parent class not required and recommended?

leezu replied:
Only needed in Python 2

leezu replied:
> Also note that, aside from the zero argument form, super() is not limited to use inside methods. The two argument form specifies the arguments exactly and makes the appropriate references. *The zero argument form only works inside a class definition, as the compiler fills in the necessary details to correctly retrieve the class being defined, as well as accessing the current instance for ordinary methods.*
https://docs.python.org/3/library/functions.html#super

mli commented:
Job PR-988/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/4/index.html

mli commented:
Job PR-988/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/5/index.html

mli commented:
Job PR-988/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/6/index.html

mli commented:
Job PR-988/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/7/index.html

fierceX commented on src/gluonnlp/model/bert.py:
```diff
@@ -17,105 +17,33 @@
-
-
-class BERTEncoder(BaseTransformerEncoder):
+class BERTEncoder(HybridBlock, Seq2SeqEncoder):
```
Why should we inherit `Seq2SeqEncoder`

leezu replied:
It designates that the `Seq2SeqEncoder` API is implemented. It's not checked anywhere in the library though, so we could also remove it

fierceX commented:
I saw that you removed some inheritance from the BERT code. I wonder if it is necessary to rewrite the BERT with almost the same code as Transformer. @leezu

leezu commented:
@fierceX I think there is a trade-off between the amount of code shared and the complexity / readability of the code. I added two more parameters two the TransformerEncoderCell and dropped BERTEncoderCell now. (Thanks for starting this discussion.)

I tend towards keeping BERTEncoder and TransformerEncoder separate.
The reason is that as the number of models published increases, it's likely that many differ in small ways and the TransformerEncoder as single shared implementation could become complex and hard to understand.

However, keeping BERTEncoder, TransformerEncoder and potential others separate, may make each of the implementations easier to understand. What do you think?

fierceX commented:
I agree with you that BERT and derived models use separate code, separated from TransformerEncoder. But this is best done separately with Transformer.py. @leezu

leezu commented:
Ok, thank you. So I suggest to merge this PR first. Potential further improvements can be done in a separate PR.

mli commented:
Job PR-988/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/11/index.html

sxjscience commented on scripts/language_model/transformer/transformer.py:
```diff
@@ -487,7 +484,7 @@ def hybrid_forward(self, F, inputs, pos_emb, mem_value, mask, segments):
 class _BaseXLNet(mx.gluon.HybridBlock):
     def __init__(self, vocab_size, num_layers=2, units=128, hidden_size=2048, num_heads=4,
-                 activation='gelu', two_stream: bool = False, scaled=True, dropout=0.0,
+                 activation='gelutanh', two_stream: bool = False, scaled=True, dropout=0.0,
```
`gelutanh` looks confusing... Use `approx_gelu`?

mli commented:
Job PR-988/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/12/index.html

mli commented:
Job PR-988/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/14/index.html

mli commented:
Job PR-988/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/15/index.html

mli commented:
Job PR-988/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/17/index.html

szha commented on src/gluonnlp/model/bert.py:
```diff
@@ -167,108 +94,155 @@ class BERTEncoder(BaseTransformerEncoder):
+            - additional_outputs of all the transformer encoder
+
+        Returns
+        -------
```
why two returns section?

leezu replied:
Taken from the `TransformerEncoder`. It's not changed in this PR
https://github.com/dmlc/gluon-nlp/pull/988/files/809187a9437466e7208ec309a608619be48b40c1#diff-411184d223687ccb09d671694e02a8c9L458-L476

leezu replied:
I didn't notice it before. Let's fix it

fierceX commented:
@leezu @eric-haibin-lin @szha 
I'd like to discuss it here. I've been watching TinyBERT lately. 
It is mentioned that an unnormalized Attention weight matrix is used. 
>In this work, the (unnormalized) attention matrix Ai is used as the fitting target instead of its softmax output softmax(Ai), since our experiments show that the former setting has a faster convergence rate and better performances.  

Is it necessary to add an option to the Attention function to return to normalized / unnormalized Attention weights

mli commented:
Job PR-988/19 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/19/index.html

szha commented:
@fierceX good suggestion. Since this PR is on the critical path for the EMNLP tutorial on Sunday, let's proceed on that separately.

mli commented:
Job PR-988/20 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/20/index.html

szha commented on src/gluonnlp/model/bert.py:
```diff
@@ -167,108 +94,147 @@ class BERTEncoder(BaseTransformerEncoder):
+        valid_length : NDArray or Symbol
+            Valid lengths of each sequence. This is usually used when part of sequence has
+            been padded. Shape (batch_size,)
+        Returns
```
newline

mli commented:
Job PR-988/21 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-988/21/index.html
